### PR TITLE
qe: Implement isolation levels for batch transactions

### PR DESCRIPTION
**Pull request by SevInf**
Add optional `isolation_level` field to `MultiDoc` and propogate it up
to connector. This mostly reuses iTx isolation level logic, implemented
on connector level.
Test-wise, we also test them in a same way iTx are tested right now:
check that valid value is accepted, invalid is rejected and rely on
quaint tests to do the right thing.

### CODE REVIEW

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/new/interactive_tx.rs [hunk 1]
```diff
@@ -174,7 +174,7 @@ mod interactive_tx {
         ];
 
         // Tx flag is not set, but it executes on an ITX.
-        runner.batch(queries, false).await?;
+        runner.batch(queries, false, None).await?;
```
Comment (garrensmith) [inline query-engine/connector-test-kit-rs/query-engine-tests/tests/new/interactive_tx.rs:177]: What would happen if a user specifies a different isolation level in a batch than what the itx is running?

Comment (garrensmith) [reply query-engine/connector-test-kit-rs/query-engine-tests/tests/new/interactive_tx.rs:177]: Do we stop a user from selecting an isolation level in a batch when it is run in an itx?

File: query-engine/core/src/executor/interpreting_executor.rs [hunk 2]
```diff
@@ -78,21 +79,21 @@ where
         &self,
         tx_id: Option<TxId>,
         operations: Vec<Operation>,
-        transactional: bool,
+        transaction: Option<BatchDocumentTransaction>,
```
Comment (garrensmith) [inline query-engine/core/src/executor/interpreting_executor.rs:82]: This is a nice improvement.

File: query-engine/core/src/executor/interpreting_executor.rs [hunk 3]
```diff
@@ -78,21 +79,21 @@ where
...
+        } else if let Some(transaction) = transaction {
             let connection_name = self.connector.name();
             let conn_span = info_span!(
                 "prisma:engine:connection",
                 user_facing = true,
                 "db.type" = connection_name.as_str()
             );
             let mut conn = self.connector.get_connection().instrument(conn_span).await?;
-            let mut tx = conn.start_transaction(None).await?;
+            let mut tx = conn.start_transaction(transaction.isolation_level).await?;
```
Comment (garrensmith) [inline query-engine/core/src/executor/interpreting_executor.rs:96]: I'm wondering if we should rather expose this as `transaction.isolation_level()` so making it a function.

File: query-engine/request-handlers/src/graphql/body.rs [hunk 4]
```diff
@@ -62,11 +67,11 @@ impl GraphQlBody {
                     .into_iter()
                     .map(|body| GraphQLProtocolAdapter::convert_query_to_operation(&body.query, body.operation_name))
                     .collect();
+                let transaction = bodies
+                    .transaction
```
Comment (garrensmith) [inline query-engine/request-handlers/src/graphql/body.rs:71]: This is a nice idea. But I find it tricky to follow and had to really think about what was happening. 
I think I would prefer:
```
 BatchDocumentTransaction::new(bodies.transaction, bodies.isolation_level)
```
That way the logic is self-contained in one place. If the transaction is `false` we ignore isolation level other wise add the `isolation` level.

**Review by garrensmith — CHANGES_REQUESTED** [review]
Just some small changes. But this looks really good. Nice work.

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/new/interactive_tx.rs (hunk 1 above)
Comment (SevInf) [reply query-engine/connector-test-kit-rs/query-engine-tests/tests/new/interactive_tx.rs:177]: In client, it's not possible: `$transaction` method is not exposed on transaction proxy.
In engine: let me check. I guess we should throw an error?

Comment (garrensmith) [reply query-engine/connector-test-kit-rs/query-engine-tests/tests/new/interactive_tx.rs:177]: I think it should be fine given you cannot hit that path via the client.

**Review by garrensmith — APPROVED** [review]
Nicely done. Thanks.

**Pull request merged**